tests/src/test/java/com/maxio/advancedbilling/controllers/invoices/InvoicesControllerReadCreditNoteTest.java

```diff
@@ -0,0 +1,265 @@
+
+    private static Product product;
+    private static Component quantityBasedComponent;
+    private static Customer customer;
+    private static Subscription subscription;
+    private static Coupon coupon;
+
+    @BeforeAll
+    static void setUp() throws IOException, ApiException {
+        ProductFamily productFamily = TEST_SETUP.createProductFamily();
+        product = TEST_SETUP.createProduct(productFamily, b -> b.priceInCents(1250));
+        customer = TEST_SETUP.createCustomer();
+        quantityBasedComponent = TEST_SETUP.createQuantityBasedComponent(productFamily.getId());
+        coupon = TEST_SETUP.createAmountCoupon(productFamily, 1250, true);
+        List<CreateSubscriptionComponent> subscriptionComponents = List.of(
+                new CreateSubscriptionComponent.Builder()
+                        .componentId(CreateSubscriptionComponentComponentId.fromNumber(quantityBasedComponent.getId()))
+                        .allocatedQuantity(CreateSubscriptionComponentAllocatedQuantity.fromNumber(50))
+                        .build());
+        subscription = TEST_SETUP.createSubscription(customer, product, s -> s
+                .couponCode(coupon.getCode())
+                .components(subscriptionComponents)
+        );
+    }
+
+    @AfterAll
+    static void teardown() throws IOException, ApiException {
+        new TestTeardown().deleteCustomer(customer);
+    }
+
+    @Test
+    void shouldReadCreditNotes() throws IOException, ApiException, InterruptedException {
+        // given
+        Integer subscriptionId = subscription.getId();
+        Invoice paidInvoice = getPaidInvoiceForSubscription(subscriptionId);
+
+        InvoicePayment payment = paidInvoice.getPayments().get(0);
+        // refund a part of the payment
+        INVOICES_CONTROLLER.refundInvoice(paidInvoice.getUid(),
+                new RefundInvoiceRequest(RefundInvoiceRequestRefund.fromRefundInvoice(
+                        new RefundInvoice.Builder()
+                                .memo("Special refund")
+                                .applyCredit(true)
+                                .paymentId(payment.getTransactionId())
+                                .amount("10.5")
+                                .build()
+                )));
+        TimeUnit.SECONDS.sleep(1);
+        List<CreditNote> creditNotes = INVOICES_CONTROLLER
+                .listCreditNotes(new ListCreditNotesInput.Builder().subscriptionId(subscriptionId).build())
+                .getCreditNotes();
+        assertThat(creditNotes).hasSize(1);
+
+        // when
+        CreditNote creditNote = INVOICES_CONTROLLER.readCreditNote(creditNotes.get(0).getUid());
+
+        // then
+        InvoiceCustomer creditNoteCustomer = creditNote.getCustomer();
+        InvoiceAddress creditNoteBillingAddress = creditNote.getBillingAddress();
+        InvoiceAddress creditNoteShippingAddress = creditNote.getShippingAddress();
+
+
+        assertAll(() -> {
```

Comment:
assertAll doesn't give any value over using plain assertions. 
If you want to print multiple failures and group some assertions logically instead of failing on the first assertion it needs to be used like this:
```
assertAll(() -> assertThat(creditNote.getUid()).isNotNull(), () -> assertThat(creditNote.getSiteId()).isNotNull()....)
```
So you can get rid of assertAll, test will work the same and there would be less indentation and block nesting.